packages/next/src/server/after/after.ts

```diff
@@ -11,36 +10,30 @@ export type AfterCallback<T = unknown> = () => T | Promise<T>
 /**
  * This function allows you to schedule callbacks to be executed after the current request finishes.
  */
-export function unstable_after<T>(task: AfterTask<T>) {
-  const callingExpression = 'unstable_after'
-
-  // TODO: This is not safe. afterContext should move to WorkStore.
-  const requestStore = requestAsyncStorage.getStore()
-  if (!requestStore) {
-    throw new Error(
-      `\`${callingExpression}\` was called outside a request scope. Read more: https://nextjs.org/docs/messages/next-dynamic-api-wrong-context`
-    )
-  }
-
-  const { afterContext } = requestStore
-  if (!afterContext) {
-    throw new Error(
-      '`unstable_after()` must be explicitly enabled by setting `experimental.after: true` in your next.config.js.'
-    )
-  }
-
+export function unstable_after<T>(task: AfterTask<T>): void {
   const workStore = workAsyncStorage.getStore()
   const cacheStore = cacheAsyncStorage.getStore()
 
   if (workStore) {
+    const { afterContext } = workStore
+    if (!afterContext) {
+      throw new Error(
+        '`unstable_after()` must be explicitly enabled by setting `experimental.after: true` in your next.config.js.'
+      )
+    }
+
+    // TODO: After should not cause dynamic.
+    const callingExpression = 'unstable_after'
     if (workStore.forceStatic) {
       throw new StaticGenBailoutError(
         `Route ${workStore.route} with \`dynamic = "force-static"\` couldn't be rendered statically because it used \`${callingExpression}\`. See more info here: https://nextjs.org/docs/app/building-your-application/rendering/static-and-dynamic#dynamic-rendering`
       )
     } else {
       markCurrentScopeAsDynamic(workStore, cacheStore, callingExpression)
     }
-  }
 
-  return afterContext.after(task)
+    afterContext.after(task)
+  } else {
+    // TODO: Error for pages?
```

Comment:
We currently prevent using `after` in pages dir at compile time (see tests [here](https://github.com/vercel/next.js/blob/d1ec45e94cdcb7a8078719abb54f1580589f49bb/test/e2e/app-dir/next-after-pages/index.test.ts#L46)) so this shouldn't ever happen. so i'd just make this an invariant for now 
```suggestion
    throw new InvariantError("Missing workStore in `unstable_after`")
```
(which will add "This is a bug in next.js" at the end etc)